File: desktop/src/runtime/intent-runtime.ts
export type IntentLayer = 'reflex' | 'behavior' | 'cognition'
export type IntentChannel = 'phase' | 'attention' | 'gesture' | 'posture' | 'expression' | 'speech'

export interface CharacterIntent<T = unknown> {
  id: string
  layer: IntentLayer
  channel: IntentChannel
  /** Omit for ambient intents that must survive conversation generation changes. */
  generation?: number
  priority: number
  createdAtMs: number
  expiresAfterMs?: number
  value: T
}

export interface ResolvedIntents {
  generation: number
  channels: Partial<Record<IntentChannel, CharacterIntent>>
}

/**
 * Engine-independent reference for the Unreal layer arbiter. It deliberately has
 * no timers or rendering dependencies, so delay/order/cancellation behavior is
 * deterministic and portable to C++ tests.
 */
export class IntentRuntime {
  private generation = 0
  private readonly intents = new Map<string, CharacterIntent>()

  currentGeneration() {
    return this.generation
  }

  beginGeneration() {
    this.generation += 1
    return this.generation
  }

  publish<T>(intent: CharacterIntent<T>) {
    if (!intent.id.trim()) throw new Error('intent id is required')
    if (!Number.isFinite(intent.priority)) throw new Error('intent priority must be finite')
    if (!Number.isFinite(intent.createdAtMs)) throw new Error('intent timestamp must be finite')
    if (intent.expiresAfterMs !== undefined && intent.expiresAfterMs < 0) {
      throw new Error('intent expiry must be non-negative')
    }
    if (intent.generation !== undefined && intent.generation < this.generation) return false
    this.intents.set(intent.id, intent as CharacterIntent)
    return true
  }

  resolve(nowMs: number): ResolvedIntents {
    const channels: ResolvedIntents['channels'] = {}
    for (const intent of this.intents.values()) {
      if (intent.generation !== undefined && intent.generation !== this.generation) continue
      if (expired(intent, nowMs)) continue
      const selected = channels[intent.channel]
      if (!selected || wins(intent, selected)) channels[intent.channel] = intent
    }
    return { generation: this.generation, channels }
  }

  compact(nowMs: number) {
    for (const [id, intent] of this.intents) {
      if (intent.generation !== undefined && intent.generation < this.generation || expired(intent, nowMs)) {
        this.intents.delete(id)
      }
    }
  }
}

function expired(intent: CharacterIntent, nowMs: number) {
  return intent.expiresAfterMs !== undefined
    && nowMs >= intent.createdAtMs + intent.expiresAfterMs
}

function wins(candidate: CharacterIntent, selected: CharacterIntent) {
  if (candidate.priority !== selected.priority) return candidate.priority > selected.priority
  if (candidate.createdAtMs !== selected.createdAtMs) return candidate.createdAtMs > selected.createdAtMs
  return candidate.id > selected.id
}

File: unreal/GahyeonStage/Source/GahyeonRuntimeCore/Private/IntentRuntime.cpp
#include "Gahyeon/IntentRuntime.h"

#include <limits>

namespace Gahyeon {

const CharacterIntent* ResolvedIntents::Find(IntentChannel channel) const {
    const auto found = Channels.find(channel);
    return found == Channels.end() ? nullptr : &found->second;
}

Generation IntentRuntime::CurrentGeneration() const {
    return generation_;
}

Generation IntentRuntime::BeginGeneration() {
    if (generation_ < std::numeric_limits<Generation>::max()) {
        ++generation_;
    }
    return generation_;
}

GenerationSyncResult IntentRuntime::SynchronizeGeneration(Generation generation) {
    if (generation < generation_) {
        return GenerationSyncResult::Stale;
    }
    if (generation == generation_) {
        return GenerationSyncResult::Unchanged;
    }
    generation_ = generation;
    return GenerationSyncResult::Advanced;
}

PublishResult IntentRuntime::Publish(CharacterIntent intent) {
    if (intent.Id.empty() || intent.ExpiresAfterMs.value_or(0) < 0) {
        return PublishResult::Invalid;
    }
    if (intent.GenerationId.has_value() && intent.GenerationId.value() < generation_) {
        return PublishResult::Stale;
    }
    intents_.insert_or_assign(intent.Id, std::move(intent));
    return PublishResult::Accepted;
}

ResolvedIntents IntentRuntime::Resolve(Millis nowMs) const {
    ResolvedIntents result;
    result.CurrentGeneration = generation_;
    for (const auto& [id, intent] : intents_) {
        if (intent.GenerationId.has_value() && intent.GenerationId.value() != generation_) {
            continue;
        }
        if (IsExpired(intent, nowMs)) {
            continue;
        }
        const auto selected = result.Channels.find(intent.Channel);
        if (selected == result.Channels.end() || Wins(intent, selected->second)) {
            result.Channels.insert_or_assign(intent.Channel, intent);
        }
    }
    return result;
}

void IntentRuntime::Compact(Millis nowMs) {
    for (auto iterator = intents_.begin(); iterator != intents_.end();) {
        const CharacterIntent& intent = iterator->second;
        const bool stale = intent.GenerationId.has_value()
            && intent.GenerationId.value() < generation_;
        if (stale || IsExpired(intent, nowMs)) {
            iterator = intents_.erase(iterator);
        } else {
            ++iterator;
        }
    }
}

bool IntentRuntime::IsExpired(const CharacterIntent& intent, Millis nowMs) {
    if (!intent.ExpiresAfterMs.has_value()) {
        return false;
    }
    const Millis duration = intent.ExpiresAfterMs.value();
    if (duration < 0) {
        return true;
    }
    if (intent.CreatedAtMs > std::numeric_limits<Millis>::max() - duration) {
        return false;
    }
    return nowMs >= intent.CreatedAtMs + duration;
}

bool IntentRuntime::Wins(
    const CharacterIntent& candidate,
    const CharacterIntent& selected) {
    if (candidate.Layer != selected.Layer) {
        return candidate.Layer == IntentLayer::Reflex
            || (candidate.Layer == IntentLayer::Behavior
                && selected.Layer == IntentLayer::Cognition);
    }
    if (candidate.Priority != selected.Priority) {
        return candidate.Priority > selected.Priority;
    }
    if (candidate.CreatedAtMs != selected.CreatedAtMs) {
        return candidate.CreatedAtMs > selected.CreatedAtMs;
    }
    return candidate.Id > selected.Id;
}

} // namespace Gahyeon
